### ISSUE TITLE
fs.readFileSync - Cannot statically evaluate fs argument

### ISSUE THREAD

**Issue by mytee306** [parcel-bundler/parcel#2363]
# 🐛 bug report

## 🎛 Configuration (.babelrc, package.json, cli command)

```js
{
  "your": { "config": "here" }
}
```

## 💻 Code Sample

https://github.com/mytee306/marked-intro

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | latest
| Node             | latest
| npm/Yarn         | latest
| Operating System | linux mint

**Comment by mytee306** [comment]
thanks for your answer, but writing 
`fs.readFileSync("./markdown.md", { encoding: "UTF-8" });`
resulted in ~
*cannot find __dirname/src/markdown.md*

**Comment by mischnic** [comment]
Does the file `src/markdown.md` exist?
If it's in the parent folder, then you need to write `../markdown.md`.

**Comment by mytee306** [comment]
it's there indeed, you may find in it [here] (https://github.com/mytee306/marked-intro/tree/master/src)

**Comment by mytee306** [comment]
@mischnic Have you been able to replicate the issue?

**Comment by mischnic** [comment]
Yes, I am.

This is the folder structure
```
src
 |- index.html
 |- index.js
 |- markdown.md
```
but inside `index.js` only `fs.readFileSync("./src/markdown.md");` works. I would expect the path to behave like `require` (relative to the current file).

**Comment by mytee306** [comment]
> Yes, I am.
> 
> This is the folder structure
> 
> ```
> src
>  |- index.html
>  |- index.js
>  |- markdown.md
> ```
> but inside `index.js` only `fs.readFileSync("./src/markdown.md");` works. I would expect the path to behave like `require` (relative to the current file).

Awesome, it worked! Although, I agree that it would make more if the *src* folder would count as the *current* folder to build a relative path off of. Just like when importing your javascript file in *index.html*.

**Comment by mischnic** [comment]
> I agree that it would make more if the src folder would count as the current folder to build a relative path off of

Yes, I would call the current behaviour a bug

**Comment by emilniklas** [comment]
Hello! I'm a first-time contributor, and I'd like to dig into this issue. Am I understanding it correctly that what's proposed here is that within a build, the behaviour should work as if `process.cwd()` was set to the directory that the entry file resides in?

So, given this structure:

```
project
 |- package.json
 |- src/
     |- index.html
     |- index.js
     |- a/
         |- a.js
     |- b/
         |- b.md
```

If `src/a/a.js` included this snippet:

```javascript
fs.readFileSync('./b/b.md');
```

And parcel was run like `parcel src/index.js`, then it would correctly find `b.md`? That would be somewhat consistent with Node in the sense that all modules share the same working directory, but it would be inconsistent when comparing these:

```shell
node src/index.js # process.cwd() == "/the/project/"
parcel src/index.js # process.cwd() == "/the/project/src"
```

The current (?) behaviour makes it work exactly like Node, in this sense:

```shell
node src/a/a.js # 🚫 "/the/project/b/b.md" not found!
parcel src/a/a.js # 🚫 "/the/project/b/b.md" not found!

cd src
node a/a.js # ✅ "/the/project/src/b/b.md" found
parcel a/a.js # ✅ "/the/project/src/b/b.md" found
```

**Comment by emilniklas** [comment]
Also, in your case @mytee306 you can do:

```javascript
const md = fs.readFileSync(require.resolve("./markdown.md"), { encoding: "UTF-8" });
```

**Comment by mischnic** [comment]
> ```
> node src/index.js # process.cwd() == "/the/project/"
> parcel src/a/a.js # 🚫 "/the/project/b/b.md" not found!
> ```
> The current (?) behaviour makes it work exactly like Node

The issue with this comparison is that `process.cwd()` doesn't exist on the web.

>```
>node src/a/a.js # 🚫 "/the/project/b/b.md" not found!
>parcel src/a/a.js # 🚫 "/the/project/b/b.md" not found!
>
>cd src
>node a/a.js # ✅ "/the/project/src/b/b.md" found
>parcel a/a.js # ✅ "/the/project/src/b/b.md" found
>```

And failing to build depending on the current directory is even worse.

**Comment by emilniklas** [comment]
> The issue with this comparison is that `process.cwd()` doesn't exist on the web.
> (`fs.readFileSync` doesn't get replaced with `--target node`)

You're correct of course, but many of us develop JavaScript for the server and the browser simultaneously, and sometimes modules are shared across those bundles. So having the same piece of JavaScript have two different behaviours in these two context (when they can work the same) seems confusing to me.

Again, the predictable, stable solution to this problem in particular is doing this:

```javascript
const md = fs.readFileSync(require.resolve("./markdown.md"), { encoding: "UTF-8" });
```

**Comment by mytee306** [comment]
> > The issue with this comparison is that `process.cwd()` doesn't exist on the web.
> > (`fs.readFileSync` doesn't get replaced with `--target node`)
> 
> You're correct of course, but many of us develop JavaScript for the server and the browser simultaneously, and sometimes modules are shared across those bundles. So having the same piece of JavaScript have two different behaviours in these two context (when they can work the same) seems confusing to me.
> 
> Again, the predictable, stable solution to this problem in particular is doing this:
> 
> ```js
> const md = fs.readFileSync(require.resolve("./markdown.md"), { encoding: "UTF-8" });
> ```

thanks for replying, you have a good point however making the above proposed changes results in 
```javascript
src/index.js:20:27: Cannot statically evaluate fs argument
  18 | });
  19 |
> 20 | const md = fs.readFileSync(require.resolve("./markdown.md"), { encoding: "UTF-8" });
     |                           ^
  21 |
  22 | document.querySelector(".app").innerHTML = marked(md);
```

[Reference](https://github.com/mytee306/marked-intro/blob/master/src/index.js)

**Comment by DeMoorJasper** [comment]
I'll change this into a feature request as `require.resolve` should probably be supported as it's pretty much what parcel currently does (without adding it explicitly).

However `process.cwd()` can not really be statically evaluated (at least not reliable).

Apologies if I've misread anything, went over this thread rather quickly.

**Comment by icopp** [comment]
@emilniklas Also see #1661 for some of my thoughts here with the "fake" `readFileSync` in a browser context.

**Comment by codeshifu** [comment]
just tell parcel you're running in a node environment with the `target` flag

e.g. `parcel app.js --target node`

**Comment by bhagyas** [comment]
@codeshifu That is all it takes!

**Comment by dreki** [comment]
@codeshifu's tip makes perfect sense and worked perfectly.

**Comment by choxinzy** [comment]
> just tell parcel you're running in a node environment with the `target` flag
> 
> e.g. `parcel app.js --target node`

can you please tell me how to do it ??

**Comment by choxinzy** [comment]
node_modules/mime/mime.js:52:32: Cannot statically evaluate fs argument
  50 |   // Read file and split into lines
  51 |   var map = {},
> 52 |       content = fs.readFileSync(file, 'ascii'),
     |                                ^
  53 |       lines = content.split(/[\r\n]+/);
  54 | 
  55 |   lines.forEach(function(line) {

how can i fix this please
